### ISSUE TITLE
Regression: special case for display boundary when suppressing autoraise

### ISSUE THREAD

**Issue by dreeves** [sbmpost/AutoRaise#67]
As I understand it, there's a safety margin to avoid accidental autoraising when resizing windows: If the mouse cursor is just barely outside of a window then we assume you didn't mean to actually leave the window, so autoraising is suppressed.

That's fine but there should be an exception to that if the mouse cursor is at the edge of the display. See [a screenshot in a previous gissue](https://github.com/sbmpost/AutoRaise/issues/8#issuecomment-743798853) for my use case here.

In short, if a window is peeking out from behind another window and that peeking-out window is flush with the display boundary, I should be able to fling the mouse cursor to the boundary to make the peeking-out window autoraise.

This used to be how it worked (after some discussion in #8) but seems to have regressed.

**Comment by sbmpost** [comment]
@dreeves 
Yes you are correct. Unfortunately it is not so much a regression of AutoRaise as it is of the latest OSX version Monterey. You can find the a comment about it on top of the AutoRaise.mm file which I will recite here:

    // It seems OSX Monterey introduced a transparent 3 pixel border around each window. This
    // means that when two windows are visually precisely connected and not overlapping, in
    // reality they are. Consequently one has to move the mouse 3 pixels further out of the
    // visual area to make the connected window raise. This new OSX 'feature' also introduces
    // unwanted raising of windows when visually connected to the top menu bar. To solve this
    // we require the mouse to be at least 5 pixels within a window boundary before raising.

**Comment by dreeves** [comment]
Oops, what I said about a window barely peeking out turns out to be a red herring! Here's my actual bug report:

## Replicata

1. Put window A in the middle of the display, nowhere near the edges of the display
2. Put window B flush against the edge of the display, eg, flush with the top of your monitor
3. Make window B big enough to overlap window A (B should stay flush with the top of the display)
4. Hover over window A so it's on top
5. Shove the mouse cursor to the top edge of the display where window B is

## Expectata

That window B autoraise, since the mouse cursor is on (the edge of) it.

## Resultata

Window B does not autoraise until I move the mouse cursor away from the edge of the screen, into the interior of window B. The mouse cursor being on the border of the window doesn't count as being on the window.

Sounds like you did that intentionally because of these 2 things:

1. One has to move the mouse 3 pixels further out of the visual area to make the connected window raise. 
2. Unwanted raising of windows when visually connected to the top menu bar.

I don't understand either of those things or if they're worse than my bug report above. I'm hoping there will turn out to be a way to have the best of all worlds!

**Comment by sbmpost** [comment]
@dreeves 
The three pixel thing is something that was introduced by OSX Monterey. If two windows are flush, in reality they overlap by three pixels. For this reason, the window under the mouse will remain the same even when you are moving the mouse outside its visual area. Because as far as OSX is concerned, the mouse only leaves the window after it is 3 pixels further outside of the visual area (yet OSX lets you drag two windows flush with the 3 pixel borders overlapping). Lets try some ascii art :)

     ___________
    | winA  |   |
    |       |___|______
    |       |   |      |
    |       |3px|      |
    |_______|___| winB |
            |___|______|

When the mouse is at the visual edge of window A, OSX requires us to move the mouse 3 pixels extra before it considers window B to be the window underneath the mouse. The same applies to the top and bottom edges which causes a problem when you have a menu opened in the top menu bar, and you want to select a menu item, but there is a window under the menu. It will make the opened menu disappear as the window is raised.

Uff difficult to explain without seeing it, but I hope it makes some sense.

It will be difficult to solve as this is some strange new OSX behavior. I will give it some more thought however.

**Comment by sbmpost** [comment]
@dreeves 
After more thought, I came up with a slightly different solution to the 3 pixel problem under OSX Monterey. I am now correcting the mouse position with 3 pixels depending on the direction it is being moved in. By doing so, flushed windows are raised as soon as the boundary is crossed. However...

1. Now when you approach a window and your mouse comes from the desktop window, it will be raised 3 pixels earlier
2. There is a horizontal border 6 pixel heigh just below the menubar where a vertical mouse move will not raise a window.
3. Other side effects?

Even with these 3 points I believe this solution to be an improvement over the previous. I am curious if you agree. To try this out, simply compile what is in this [branch](https://github.com/sbmpost/AutoRaise/tree/67-take-into-account-3px-border)

**Comment by dreeves** [comment]
This solves my problem! I still don't understand the tradeoffs with the menubar and stuff but I am very happy now that I can autoraise windows that are flush with the edge of my monitor. Thank you so much for doing this! I hope this makes it to master but I'll happily stay on this branch in the meantime.

**Comment by sbmpost** [comment]
@dreeves 
I think this might make it to master but I would like to keep this issue open for a while in case you have more feedback after a few days of use ;-)

Thanks for reporting as always.

**Comment by sbmpost** [comment]
@dreeves 
Is this branch still working properly for you? If yes, then we can merge this to master ;-)

**Comment by sbmpost** [comment]
@dreeves 
The new 3px solution for Monterey was merged to master so I am closing this issue.
